contracts/governance/Governor.sol

```diff
@@ -40,12 +42,19 @@ abstract contract Governor is Context, ERC165, EIP712, IGovernor {
 
     mapping(uint256 => ProposalCore) private _proposals;
 
+    // This mapping keeps track of the governor operating on itself. Calls to functions protected by the
+    // {onlyGovernance} modifier needs to be whitelisted in this mapping. Whitelisting is set in {_beforeExecute},
+    // and reset in {_afterExecute}. This ensures that the execution of {onlyGovernance} protected calls can only
+    // be achieved through successful proposals.
+    mapping(bytes => Counters.Counter) private _governanceCall;
```

Comment:
when pushing into a `bytes32[]`, you pay one sstore to update the length, in addition to the sstore to set the counter.

Reply:
I've added private functions to add, use and clear the permissions.

Reply:
It's true that the current implementation is very simple. I think it's worth comparing the costs though. I'm not sure which of the two options will be cheaper, and by how much. The main motivation behind my proposal is to avoid the for loop in afterExecute.

Reply:
instead of execution the for loop once in afterExecute, you'll execute it N time in onlyGovernance(), no?

Reply:
It's smaller N and in a sense you only pay for what you use. Where in the current implementation there is N overhead regardless of whether you use governance actions.

Reply:
I don't understand where the array is more effective. For n calls, and m governance calls (m <= n)

With a mapping:
- setting: O(n)
- checking: m times O(1) → O(m)
- resetting: O(n)

With an array:
- setting: O(n) with a bigger constant (2 slots written instead of one)
- checking: m times O(m) → O(m²)
- resetting: O(1)

With an array, the 2 slots per set are equivalent to set+reset the mapping... and the checking is more expensive with a mapping. So mathematically, I don't see the array being obviously better. Maybe experimenting would prove me wrong, but we would need to see actual numbers.

Reply:
Setting can be optimized so that the length is written to storage only once at the end of the loop. This introduces some code complexity but the optization makes sense.

It's true that there are O(m²) reads but my argument is that m is smaller than n.

Overall I can imagine the constant being smaller. But we'd need to test it.

Reply:
Actually, we can assume that the batch is executed sequentially. Then the array would be used exactly like a FIFO queue so linear search is not even necessary!

By the way I think this is an interesting discussion to get inspiration for data structures. A FIFO queue could make a lot of sense. We can pack start index and length as two uint128 in the same storage slot. This sort of storage optimization is the kind of thing that makes data structure libraries very valuable.

Reply:
https://gist.github.com/Amxx/9d75936806ea77a801453b56c92c2229

Reply:
At that point, I think we need more time to study different options. I would not rush that to make it to 4.5, and I would not delay 4.5 to wait for this.

Reply:
Let's think about it a bit. I'd be ok with postponing this improvement to the next release if we can't quickly come to a conclusion.